### PR TITLE
[receiver/azureeventhub] add metric support

### PR DESCRIPTION
**Pull request by cparkins**
**Description:** <Describe what has changed.>
Add support to the Azure Event Hub receiver to accept Metrics from Azure Resources and convert them into OpenTelemetry Metrics.

**Link to tracking Issue:** <Issue number if applicable>
[18690]

**Testing:** <Describe what testing was performed and which tests were added.>
Tested locally with data from an Event Hub Instance with Azure Diagnostic Metrics configured to stream there.

**Documentation:** <Describe the documentation added.>
Added documentation to describe the approach taken to mapping the data from Azure.

### CODE REVIEW

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler.go [hunk 1]
```diff
@@ -69,10 +80,75 @@ type azureLogRecord struct {
...
+	}
+
+	return l, nil
+}
+
 // asTimestamp will parse an ISO8601 string into an OpenTelemetry
 // nanosecond timestamp. If the string cannot be parsed, it will
 // return zero and the error.
 func asTimestamp(s string) (pcommon.Timestamp, error) {
+
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler.go:151]:
```suggestion
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler.go [hunk 2]
```diff
@@ -84,6 +160,7 @@ func asTimestamp(s string) (pcommon.Timestamp, error) {
 // OpenTelemetry severity numbers. If the log level is not
 // valid, then the 'Unspecified' value is returned.
 func asSeverity(s string) plog.SeverityNumber {
+
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler.go:163]:
```suggestion
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler.go [hunk 3]
```diff
@@ -98,21 +175,13 @@ func asSeverity(s string) plog.SeverityNumber {
...
-	}
-}
-
 // extractRawAttributes creates a raw attribute map and
 // inserts attributes from the Azure log record. Optional
 // attributes are only inserted if they are defined. The
 // azureDuration value is only set if the value in the
 // Azure log record can be parsed as an integer.
 func extractRawAttributes(log azureLogRecord) map[string]interface{} {
+
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler.go:184]:
```suggestion
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go [hunk 4]
```diff
@@ -75,7 +77,7 @@ var maximumLogRecord = func() plog.LogRecord {
 	return lr
 }()
 
-func TestAsTimestamp(t *testing.T) {
+func Test_AsTimestamp(t *testing.T) {
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go:80]: These changes just add to the diff, mind doing that as a chore later?
```suggestion
func TestAsTimestamp(t *testing.T) {
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go [hunk 5]
```diff
@@ -87,7 +89,7 @@ func TestAsTimestamp(t *testing.T) {
 	assert.Equal(t, pcommon.Timestamp(0), nanos)
 }
 
-func TestAsSeverity(t *testing.T) {
+func Test_AsSeverity(t *testing.T) {
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go:92]:
```suggestion
func TestAsSeverity(t *testing.T) {
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go [hunk 6]
```diff
@@ -103,7 +105,7 @@ func TestAsSeverity(t *testing.T) {
 	}
 }
 
-func TestSetIf(t *testing.T) {
+func Test_SetIf(t *testing.T) {
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go:108]:
```suggestion
func TestSetIf(t *testing.T) {
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go [hunk 7]
```diff
@@ -124,7 +126,7 @@ func TestSetIf(t *testing.T) {
 	assert.Equal(t, "ok", actual)
 }
 
-func TestExtractRawAttributes(t *testing.T) {
+func Test_ExtractRawAttributes(t *testing.T) {
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go:129]:
```suggestion
func TestExtractRawAttributes(t *testing.T) {
```

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go [hunk 8]
```diff
@@ -229,7 +231,7 @@ func TestExtractRawAttributes(t *testing.T) {
 
 }
 
-func TestDecodeAzureLogRecord(t *testing.T) {
+func Test_UnmarshalLogs(t *testing.T) {
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go:234]:
```suggestion
func TestUnmarshalLogs(t *testing.T) {
```

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	err := decoder.Decode(&azureMetrics)
+	if err != nil {
+		return md, err
+	}
+
+	resourceMetrics := md.ResourceMetrics().AppendEmpty()
+	scopeMetrics := resourceMetrics.ScopeMetrics().AppendEmpty()
+
+	scopeMetrics.Scope().SetName(receiverScopeName)
+	scopeMetrics.Scope().SetVersion(r.buildInfo.Version)
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:85]: could this go at the resource level?

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(nanos.AsTime().Add(-time.Minute)))
+		} else {
+			r.logger.Warn("Unhandled Time Grain")
+		}
+		dp.SetTimestamp(nanos)
+
+		// The Azure resource ID will be pulled into a common resource attribute.
+		// This implementation assumes that a single metric message from Azure will
+		// contain ONLY metrics from a single resource.
+		if resourceID != "" {
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:120]: please move out of the for loop

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+	metrics := scopeMetrics.Metrics()
+	metrics.EnsureCapacity(len(azureMetrics.Records))
+
+	resourceID := ""
+	for _, azureMetric := range azureMetrics.Records {
+		resourceID = azureMetric.ResourceID
+		nanos, err := asTimestamp(azureMetric.Time)
+		if err != nil {
+			continue
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:96]: worth logging something?

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+		// }
+
+		dp := metric.SetEmptySummary().DataPoints().AppendEmpty()
+		dp.SetCount(uint64(azureMetric.Count))
+		dp.SetSum(azureMetric.Total)
+
+		if azureMetric.TimeGrain == "PT1M" {
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(nanos.AsTime().Add(-time.Minute)))
+		} else {
+			r.logger.Warn("Unhandled Time Grain")
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:113]: can you warn and print the grain?

File: receiver/azureeventhubreceiver/eventhubhandler_test.go [hunk 13]
```diff
@@ -22,11 +22,14 @@ import (
...
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer"
 	"go.opentelemetry.io/collector/consumer/consumertest"
 	"go.opentelemetry.io/collector/obsreport"
 	"go.opentelemetry.io/collector/receiver/receivertest"
+	"go.uber.org/zap"
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/eventhubhandler_test.go:32]: you need to run goimports and will need to move this import into the group up.

**Review by atoulme — COMMENTED** [review]
Impressive work! We can talk later about using the shared component library to reuse the receiver between metrics and logs. I have some nits.

File: receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go (hunk 4 above)
Comment (cparkins) [reply receiver/azureeventhubreceiver/azureresourcelogs_unmarshaler_test.go:80]: No problem, I was just looking at tests in other receivers. I use the kafka and azureblob receiver as examples a lot.

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+		}
+
+		metric := metrics.AppendEmpty()
+		metric.SetName(azureMetric.MetricName)
+
+		// for i := desc.attrs.Iter(); i.Next(); {
+		// 	dp.Attributes().PutStr(string(i.Attribute().Key), i.Attribute().Value.AsString())
+		// }
+
+		dp := metric.SetEmptySummary().DataPoints().AppendEmpty()
```
Comment (cparkins) [inline receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:106]: I have some concerns about using the Summary type. I think a Gauge may be more appropriate, so I will likely make that change today.

**Comment by cparkins** [comment]
> Impressive work! We can talk later about using the shared component library to reuse the receiver between metrics and logs. I have some nits.

I had code to use the shared components but thought I had removed it completely. I can add it back. I'm not sure if it's appropriate or not as I don't completely understand it's purpose.

**Comment by atoulme** [comment]
Exactly, no worries and it’s out of scope. Let’s deliver this first.

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go (hunk 14 above)
Comment (atoulme) [reply receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:106]: No opinion from me here. You have more context.

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go (hunk 12 above)
Comment (atoulme) [reply receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:113]: Just use an additional argument (from memory)
```suggestion
			r.logger.Warn("Unhandled Time Grain", zap.Error(err))
```

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go (hunk 9 above)
Comment (atoulme) [reply receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:85]: The only difference is economical: the attributes apply anyway, it’s just that the resource is more generic.

Comment (cparkins) [reply receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:86]: I could use some help in getting this change in place. I looked for some examples but couldn't determine how to set the values on the Resource unless I use Attributes.

File: receiver/azureeventhubreceiver/eventhubhandler_test.go (hunk 13 above)
Comment (cparkins) [reply receiver/azureeventhubreceiver/eventhubhandler_test.go:32]: Looking at the go.mod it appears that zap was available.  Is there something I'm missing?

File: receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go (hunk 9 above)
Comment (atoulme) [reply receiver/azureeventhubreceiver/azureresourcemetrics_unmarshaler.go:86]: Sorry, I am misleading you on this one. You are correct that name and version are set at the scope level.

File: receiver/azureeventhubreceiver/eventhubhandler_test.go (hunk 13 above)
Comment (atoulme) [reply receiver/azureeventhubreceiver/eventhubhandler_test.go:32]: it's just the way the imports are set. Try to run `make lint` or `make impi` in the folder to see the errors.

**Comment by cparkins** [comment]
For future reference the Azure Monitor Scraper Receiver may be a good place to look for improvements to the Metrics implementation: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18898

With the changes proposed there the Azure API is queried for each Resource to determine the Metrics that are valid.  There is documentation from Microsoft that could be used to build the current Metrics per Resource Type: https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/azure-monitor/essentials/metrics-supported.md.  But the preferred route would be to query Azure Monitor: https://learn.microsoft.com/en-us/rest/api/monitor/metric-definitions/list?source=recommendations&tabs=HTTP.

**Review by atoulme — APPROVED** [review]

**Pull request merged**